### PR TITLE
RATEST-74: Failing Test: ignored AddFormTests

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/RATEST-74
 cc @kaweesi  kindly review thanks

### CODE REVIEW

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 1]
```diff
@@ -18,22 +16,40 @@
  */
 public class AddFormTest extends TestBase {
     private HomePage homePage;
+    private static String name = "newFormTest";
+    private static String description = "description of new form";
+    private static  String version = "1.2";
+
+    private static final By NEW_HTML_FORM = By.cssSelector("a[href='htmlForm.form']");
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:23]: do you need this field here? and if yes, can u make use of https://github.com/openmrs/openmrs-distro-referenceapplication/pull/360/files#diff-268be05503616ccdc23bf47f929298bcf3eb451ca81e8b2e1d0b9dd90ca0315dR8

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 2]
```diff
@@ -18,22 +16,40 @@
...
         patientDashboardPage = new ClinicianFacingPatientDashboardPage(page);
+
     }
 
-    @Ignore //ignore due to moving forms functionality
     @Test
     public void addFormTest() throws Exception {
+     manageHtmlFormsPage =  homePage.goToAdministration().clickOnManageHtmlForms();
+
+     if (manageHtmlFormsPage.getElementsIfExisting(By.xpath("//*[contains(text(), '"+ name+ "')]")).isEmpty()){
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:47]: try rightly formatting your code to well format `+ name+ `

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageFormsPage.java [hunk 3]
```diff
@@ -55,7 +55,9 @@ public void addIcon(String icon) {
 
     public boolean addPresent() {
         try {
-            return driver.findElement(ADD) != null;
+            
+            boolean present = driver.findElement(ADD) != null;
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/ManageFormsPage.java:59]: u can revert this change

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 4]
```diff
@@ -18,22 +16,40 @@
...
+         manageHtmlFormsPage.clickOnNewHtmlForm();
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:48]: can you make clickOnNewHtmlForm to return the new page below this line

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 5]
```diff
@@ -18,22 +16,40 @@
...
+         HtmlFormsPage newHtmlForms = new HtmlFormsPage(manageHtmlFormsPage);
+         newHtmlForms.CreateNewFormTest(name,description,version);
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:50]: you can some wrong naming here with the starting character of `CreateNewFormTest` method, https://wiki.openmrs.org/display/docs/Java+Conventions

File: ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java [hunk 6]
```diff
@@ -21,6 +21,8 @@
     private final static By MANAGE_VISIT_TYPES = By.cssSelector("#legacyui-manageVisitTypes a");
     private static final By MANAGE_PROVIDERS = By.cssSelector("#content a[href='/openmrs/admin/provider/index.htm']");
     private static final By MANAGE_PERSONS = By.cssSelector("#content a[href=\"/openmrs/admin/person/index.htm\"]");
+    private static final By MANAGE_HTMLFORMS_PAGE = By.cssSelector("#content a[href*=\"/module/htmlformentry/htmlForms.list\"]");
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java:24]: instead of escaping double quotes, replace with single quotes to make this more readable

**Review by kaweesi — COMMENTED** [review]
This is a great progress @sherrif10 just check my few remaining comments and we will be ready soon

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java (hunk 1 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:23]: This line of code is looks similar with the above code  , unless am getting you wrongly

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java (hunk 4 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:48]: i think we had agreed on this that it returns htmlFormsPage, which is hard coded on the below line

File: ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java (hunk 6 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java (hunk 5 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java (hunk 2 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/ManageFormsPage.java (hunk 3 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/ManageFormsPage.java]: Resolved thanks

**Comment by sherrif10** [comment]
Thanks alot for your review, i have answered your review  thanks

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java (hunk 4 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:44]: Actually this method returns HtmlFormsPage

**Comment by sherrif10** [comment]
Any more suggestion on this

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 7]
```diff
@@ -18,22 +16,36 @@
...
-        assertPage(homePage);
         headerPage = new HeaderPage(driver);
+        administrationPage = new AdministrationPage(page);
         manageForm = new ManageFormsPage(driver);
         patientDashboardPage = new ClinicianFacingPatientDashboardPage(page);
+
     }
 
-    @Ignore //ignore due to moving forms functionality
     @Test
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:39]: you need @Category(BuildTests.class) on each Test

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:41]: Resolved thanks

File: ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java [hunk 8]
```diff
@@ -21,6 +21,8 @@
     private final static By MANAGE_VISIT_TYPES = By.cssSelector("#legacyui-manageVisitTypes a");
     private static final By MANAGE_PROVIDERS = By.cssSelector("#content a[href='/openmrs/admin/provider/index.htm']");
     private static final By MANAGE_PERSONS = By.cssSelector("#content a[href=\"/openmrs/admin/person/index.htm\"]");
+    private static final By MANAGE_HTMLFORMS_PAGE = By.cssSelector("#content a[href*=\'/module/htmlformentry/htmlForms.list\']");
```
Comment (kaweesi) [inline ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java:24]: no need to escape ' and replace all \" with '

Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java:24]: Resolved thanks

**Comment by sherrif10** [comment]
cc @kaweesi  ,is this ready to get merged

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 9]
```diff
@@ -18,22 +18,37 @@
...
+    @Category(BuildTests.class)
     public void addFormTest() throws Exception {
+     manageHtmlFormsPage =  homePage.goToAdministration().clickOnManageHtmlForms();
+
+     if (manageHtmlFormsPage.getElementsIfExisting(By.xpath("//*[contains(text(), '"+ name + "')]")).isEmpty()){
```
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:46]: refactor` manageHtmlFormsPage.getElementsIfExisting(By.xpath("//*[contains(text(), '"+ name + "')]")` into a boolean method in manageHtmlFormsPage and use it here instead

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java [hunk 10]
```diff
@@ -18,22 +18,37 @@
...
+         manageHtmlFormsPage.clickOnNewHtmlForm();
+         HtmlFormsPage newHtmlForms = new HtmlFormsPage(manageHtmlFormsPage);
+         newHtmlForms.createNewFormTest(name,description,version);
```
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:49]: invoke this method directly on the above line and get rid of the `newHtmlForms` variable

File: ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java [hunk 11]
```diff
@@ -20,7 +20,9 @@
     private static final By MANAGE_MODULES = By.cssSelector("#legacyui-manageModules a");
     private final static By MANAGE_VISIT_TYPES = By.cssSelector("#legacyui-manageVisitTypes a");
     private static final By MANAGE_PROVIDERS = By.cssSelector("#content a[href='/openmrs/admin/provider/index.htm']");
-    private static final By MANAGE_PERSONS = By.cssSelector("#content a[href=\"/openmrs/admin/person/index.htm\"]");
+    private static final By MANAGE_PERSONS = By.cssSelector("#content a[href=\'/openmrs/admin/person/index.htm\']");
```
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/AdministrationPage.java:23]: you don't need to escape `'`

File: ui-tests/src/test/java/org/openmrs/reference/page/HtmlFormsPage.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.openmrs.uitestframework.page.Page;
+import org.openqa.selenium.By;
+
+public class HtmlFormsPage extends Page {
+
+    private static final By NEW_HTML_FORM = By.cssSelector("a[href='htmlForm.form']");
+    private static final By NAME = By.name("form.name");
+    private static  By DESCRIPTION = By.name("form.description");
+    private static  By SAVE_FORM = By.cssSelector("input[type=\"submit\"]");
```
Comment (k-joseph) [inline ui-tests/src/test/java/org/openmrs/reference/page/HtmlFormsPage.java:11]: avoid escaping which makes things not readable enough, replace those with single quotes across the PR

File: ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java (hunk 10 above)
Comment (sherrif10) [reply ui-tests/src/test/java/org/openmrs/reference/AddFormTest.java:49]: This will  cause null ponter exception and hence will need me to handle which will look not more readable

**Comment by sherrif10** [comment]
Closing this and Reopenning [Ticket](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/396)
